core/src/ops/matmul/mir_quant.rs

```diff
@@ -731,149 +802,330 @@ mod test {
             b_scale: 1.0,
             c_scale: 1.0,
         }
-        .check()
-    }
-
-    #[derive(Debug)]
-    struct QMatMulProblem {
-        a: Array2<i8>,
-        b: Array2<i8>,
-        bias: Tensor,
-        a0: i8,
-        b0: i8,
-        c0: i8,
-        a_scale: f32,
-        b_scale: f32,
-        c_scale: f32,
+        .check();
     }
 
     fn round_ties_to_right(x: f32) -> i32 {
         (x + 0.5).floor() as i32
     }
 
-    impl QMatMulProblem {
-        fn check(&self) {
-            let r = self.reference();
-            let t = self.tract();
-            if r.iter().zip(t.iter()).any(|(r, t)| r.max(t) - r.min(t) > 1) {
-                panic!("mismatch! refernce: {:?} tract: {:?}", r, t)
+    fn scale() -> BoxedStrategy<f32> {
+        prop_oneof![Just(1.0), (1i32..=20).prop_map(|x| x as f32 / 20.0)].boxed()
+    }
+
+    macro_rules! impl_qmmp {
```

Comment:
Mostly curious... can we not achieve this with type parameters on QProblem instead of a macro ?

Reply:
i tried that in the beginning and was running into trait bound issues, iirc not so much for `Datum` but more for `Arbitrary`. it might be possible to express this without a macro now in the latest iteration of the tests, but i haven't tried that.